### PR TITLE
Refactor some of the skylark rules.

### PR DESCRIPTION
**Pull request by DrMarcII**
-  Use a web_test provider with a struct value instead of
   of multiple separate providers.
-  Remove shared.bzl and move functionality into smaller
   more focused bazel libraries.
-  Consistently use namespaces for bazel libraries.
-  Make other modifications to improve the consistency of
   the project.
-  Use sets where appropriate.

### CODE REVIEW

File: web/internal/metadata.bzl [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      return files earlier in the result.
+  Returns:
+    A sequence of all of the web_test_metadata files in the specificied 
+    attributes.
+  """
+  metadata_files = set()
+
+  for attr_name in reversed(attr_names):
+    attr = getattr(ctx.attr, attr_name)
+    if lists.is_list_like(attr):
```
Comment (jart) [inline web/internal/metadata.bzl]: There's a lot of clever reflection going on in these methods that might not be necessary.

While working on cl/137763822 with @tituswinters a few days ago, he shared with me go/totw/60 which argues why we should reconsider writing clever code. It's a worthwhile read, but another thing to consider is this: I understand how to use the Skylark APIs, but I don't understand how to follow the behavior of this framework built on top of it.

Maybe in a follow up PR, you'll consider refactoring the code to be more unfancy and less frameworky?

File: web/internal/metadata.bzl [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  Args:
+    ctx: a Skylark rule context.
+    output: File object. The file to write the metadata to.
+    environment: string. The Web Test Launcher environment name.
+    browser_label: Label. The label for a browser rule.
+    test_label: Label. The label for the test being executed.
+    web_test_files: sequence of web_test_file structs. The named files needed
+      for this configuration.
+  """
+  content = '{\n  "_comment": "generated file for %s"' % ctx.label
```
Comment (DrMarcII) [inline web/internal/metadata.bzl]: Switch to to_json. Will consider trade-offs of using a proto.

File: web/internal/metadata.bzl (hunk 1 above)
Comment (DrMarcII) [inline web/internal/metadata.bzl]: I figured out that what I was really doing wrong was using some values in data like deps (allowing them to affect how things are built). I have added a deps attribute to several rules that requires "web_test" providers. I can now merge these into the metadata without worrying about checking for the provider, and am no longer concerned about the weird possible (mis)uses that led me to experiment with aspects in that other PR.

File: web/web.bzl [hunk 3]
```diff
@@ -171,18 +169,33 @@ def browser(testonly=True, **kwargs):
...
   """Wrapper around web_test to correctly set defaults.
   
   Args:
     data
     size: default = "large"
     **kwargs: see web_test in //web/internal:web_test.bzl
   """
+  config = config or str(Label("//web:default_config"))
+  launcher = launcher or str(Label("//go/launcher:main"))
   data = lists.clone(data)
```
Comment (jart) [inline web/web.bzl]: Rather than doing this, consider doing the following in web_test.bzl:

```python
  return struct(
      # ...
      runfiles=ctx.runfiles(
          files=(ctx.files.browser +
                 ctx.files.config +
                 ctx.files.launcher +
                 ctx.files.test),
          transitive_files=(collect_default_runfiles([ctx.attr.browser,
                                                      ctx.attr.config,
                                                      ctx.attr.launcher,
                                                      ctx.attr.test]) |
                            collect_data_runfiles([ctx.attr.browser,
                                                   ctx.attr.config,
                                                   ctx.attr.launcher,
                                                   ctx.attr.test])),
          collect_data=True,
          collect_default=True))

def collect_default_runfiles(targets):
  """Aggregates default runfiles from targets."""
  data = set()
  for target in targets:
    data += _get_runfiles(target, "default_runfiles")
  return data

def collect_data_runfiles(targets):
  """Aggregates data runfiles from targets."""
  data = set()
  for target in targets:
    data += _get_runfiles(target, "runfiles")
    data += _get_runfiles(target, "data_runfiles")
  return data

def _get_runfiles(target, attribute):
  runfiles = getattr(target, attribute, None)
  if runfiles:
    # Legacy rules create gigantic .jslib files we don't need.
    return [f for f in runfiles.files if not f.path.endswith(".jslib")]
  return []
```

Please note that I'm not 100% sure this is what you want to do. I'm actually talking with Laurent (Master of Skylark) in cl/138992178 hoping he's going to help me grok the runfiles thing, because this review made me realize that I was doing it wrong too.

Comment (DrMarcII) [inline web/web.bzl]: That is similar to what I was doing before, but I found that it doesn't work from some rule types (specifically I remember that when test referred to a py_test it didn't work because the py runfiles has an empty file provider for __init__.py files that gets missed by this).

Comment (jart) [inline web/web.bzl]: That's an interesting edge case. I'll need to bring that up with Laurent.
